### PR TITLE
feat(discussionPromptQuestion): Generate discussion prompt question for discussions

### PR DESCRIPTION
**Pull request by tianrunhe**
# Description
AI-generated discussion prompt questions, gated by the feature flag `AIGeneratedDiscussionPrompt`. This will start from **internal testing**: after it's released, will manually add the feature flag.

## Demo
Screenshot 2023-07-09 at 11 22 41

## Testing scenarios
- [ ] 
  - In a retro meeting, write some reflections
  - Go to the discussion phase and see the prompt question

## Final checklist

- [x] I checked the [code review guidelines](../docs/codeReview.md)
- [x] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [x] I have performed a self-review of my code, the same way I'd do it for any other team member
- [x] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [x] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [x] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [x] PR title is human readable and could be used in changelog

### CODE REVIEW

File: packages/server/utils/OpenAIServerManager.ts [hunk 1]
```diff
@@ -41,6 +42,64 @@ class OpenAIServerManager {
...
+    ${reflections
+      .map(({plaintextContent}) => plaintextContent.trim().replace(/\n/g, '\t'))
+      .join('\n')}`
+    try {
+      const response = await this.openAIApi.createChatCompletion({
+        model: 'gpt-3.5-turbo',
+        messages: [
+          {
+            role: 'user',
+            content: prompt
```
Comment (nickoferrall) [inline packages/server/utils/OpenAIServerManager.ts:78]: I copied and pasted the reflections from our last Squad retro and gave it a spin, and I think the prompts are generally very good! 

I did get one that repeated the instructions though:
Screenshot 2023-07-10 at 20 00 05

And two of them included quotation marks around the discussion topic:

Screenshot 2023-07-10 at 20 00 32

I think that's fine for internal testing though. The prompt can always be updated.

**Review by nickoferrall — CHANGES_REQUESTED** [review]
Very cool feature! Looking forward to seeing this in a Squad retro soon

File: packages/server/graphql/mutations/helpers/generateGroupSummaries.ts [hunk 2]
```diff
@@ -37,8 +37,20 @@ const generateGroupSummaries = async (
         ({plaintextContent}) => plaintextContent
       )
       const summary = await manager.getSummary(reflectionTextByGroupId)
-      if (!summary) return
-      return r.table('RetroReflectionGroup').get(group.id).update({summary}).run()
+      const discussionPromptQuestion = facilitator.featureFlags.includes(
```
[the remark this answers is not in the document]
Comment (tianrunhe) [reply packages/server/graphql/mutations/helpers/generateGroupSummaries.ts:40]: With our recent discussion, I think future road leads to a single `enable AI` toggle that controls all AI related features, I keep it as-is. I think if this new feature passes internal test and is ready to release to our customers, we can then talk about refactoring the logic & renaming. WDYT?

File: packages/server/graphql/mutations/helpers/addSummariesToThreads.ts [hunk 3]
```diff
@@ -20,21 +20,39 @@ const addSummariesToThreads = async (
   const {tier} = team
   const commentPromises = stages.map(async (stage, idx) => {
     const group = groups.find((group) => group.id === stage.reflectionGroupId)
-    if (!group?.summary) return
-    const explainerText = tier === 'starter' ? AIExplainer.STARTER : AIExplainer.PREMIUM_REFLECTIONS
-    const html =
+    if (!group?.summary && !group?.discussionPromptQuestion) return
```
[the remark this answers is not in the document]
Comment (tianrunhe) [reply packages/server/graphql/mutations/helpers/addSummariesToThreads.ts:23]: Didn't change for the reason explained [below](https://github.com/ParabolInc/parabol/pull/8462#discussion_r1264263809). Also, prompt question is sort of related to a summary of the thread topic? LOL I'm coming up with excuses to be lazy probably.

File: packages/server/utils/OpenAIServerManager.ts (hunk 1 above)
Comment (tianrunhe) [reply packages/server/utils/OpenAIServerManager.ts:78]: Made some improvements

File: packages/server/utils/OpenAIServerManager.ts [hunk 4]
```diff
@@ -78,18 +78,15 @@ class OpenAIServerManager {
             content: prompt
           }
         ],
-        temperature: 0.7,
-        max_tokens: 80,
-        top_p: 1,
-        frequency_penalty: 0,
-        presence_penalty: 0
+        temperature: 0.3,
```
Comment (tianrunhe) [inline packages/server/utils/OpenAIServerManager.ts:82]: To make the output less random

File: packages/server/utils/OpenAIServerManager.ts [hunk 5]
```diff
@@ -78,18 +78,15 @@ class OpenAIServerManager {
             content: prompt
           }
         ],
-        temperature: 0.7,
-        max_tokens: 80,
-        top_p: 1,
```
Comment (tianrunhe) [inline packages/server/utils/OpenAIServerManager.ts]: According to OpenAI [doc](https://platform.openai.com/docs/api-reference/chat/create), best practice is to not set **both** `temperature` and `top_p`

File: packages/server/utils/OpenAIServerManager.ts [hunk 6]
```diff
@@ -78,18 +78,15 @@ class OpenAIServerManager {
...
-        frequency_penalty: 0,
-        presence_penalty: 0
```
Comment (tianrunhe) [inline packages/server/utils/OpenAIServerManager.ts]: 0 are default values so removing them

File: packages/server/graphql/mutations/helpers/addSummariesToThreads.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/addSummariesToThreads.ts:23]: 😄 if I read `addSummariesToThread`, I would just expect summaries to be in there. I think we should rename it if prompts will also be here or it'll be confusing

**Review by nickoferrall — APPROVED** [review]
LGTM! I think the file needs renaming, but other than that, I'm looking forward to seeing this in a Parabol Retro soom!

**Review by nickoferrall — APPROVED** [review]
LGTM! I think the file needs renaming, but other than that, I'm looking forward to seeing this in a Parabol Retro soon!

File: packages/server/graphql/mutations/helpers/addAIGeneratedContentToThreads.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+) => {
+  const [r, groups, team] = await Promise.all([
+    getRethink(),
+    dataLoader.get('retroReflectionGroupsByMeetingId').load(meetingId),
+    dataLoader.get('teams').loadNonNull(teamId)
+  ])
+  const {tier} = team
+  const commentPromises = stages.map(async (stage, idx) => {
+    const group = groups.find((group) => group.id === stage.reflectionGroupId)
+    if (!group?.summary && !group?.discussionPromptQuestion) return
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/helpers/addAIGeneratedContentToThreads.ts:23]: -1 If only one of those is not defined, we will still add both comments, one of them will just be undefined

File: packages/server/graphql/mutations/helpers/generateGroupSummaries.ts [hunk 8]
```diff
@@ -39,11 +39,23 @@ const generateGroupSummaries = async (
         ({plaintextContent}) => plaintextContent
       )
       const fullSummary = await manager.getSummary(reflectionTextByGroupId)
-      if (!fullSummary) return
-      const summary = fullSummary.slice(0, 2000)
+      const fullQuestion = facilitator.featureFlags.includes('AIGeneratedDiscussionPrompt')
+        ? await manager.getDiscussionPromptQuestion(group.title ?? 'Unknown', reflectionsByGroupId)
+        : undefined
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/helpers/generateGroupSummaries.ts:44]: +1 Any reason why we don't await both at the same time in `Promise.all`?

File: packages/server/graphql/mutations/resetRetroMeetingToGroupStage.ts [hunk 9]
```diff
@@ -112,13 +112,13 @@ const resetRetroMeetingToGroupStage = {
...
         .updateTable('RetroReflectionGroup')
-        .set({voterIds: []})
+        .set({voterIds: [], summary: null, discussionPromptQuestion: null})
         .where('id', 'in', reflectionGroupIds)
         .execute(),
       r
         .table('RetroReflectionGroup')
         .getAll(r.args(reflectionGroupIds))
-        .update({voterIds: []})
+        .update({voterIds: [], summary: '', discussionPromptQuestion: ''})
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/resetRetroMeetingToGroupStage.ts:121]: -1 Why don't we set the summary and question to `null` here? Seems like it's introducing an inconsistency between R and PG for no reason.

**Review by Dschoordsch — CHANGES_REQUESTED** [review]

File: packages/server/graphql/mutations/helpers/addAIGeneratedContentToThreads.ts [hunk 10]
```diff
@@ -18,41 +32,31 @@ const addAIGeneratedContentToThreads = async (
...
-        : `<html><body><p><b>🤖 Discussion Question</b></p><p>${group.discussionPromptQuestion}</p></body></html>`
-    const discussionPromptQuestionBlock = convertHtmlToTaskContent(discussionPromptQuestionHtml)
-    const discussionPromptQuestionCommentInput = {
-      discussionId: stage.discussionId,
-      content: discussionPromptQuestionBlock,
-      threadSortOrder: 1, // make sure it's after the topic summary
-      createdBy: PARABOL_AI_USER_ID
+    if (group.summary) {
+      const topicSummaryExplainerText =
+        tier === 'starter' && idx === 0 ? AIExplainer.STARTER : AIExplainer.PREMIUM_REFLECTIONS
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/helpers/addAIGeneratedContentToThreads.ts:42]: -1 Now we're showing the explainer each time, but the first stage sees the starter one for starter
```ts
idx === 0 ? (tier === 'starter' ? AIExplainer.STARTER : AIExplainer.PREMIUM_REFLECTIONS) : null
```

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
I did not see the question yet, but I assume you've tested this properly.

**Review by Dschoordsch — APPROVED** [review]

**Pull request merged**